docs/_includes/Statements.md

```diff
@@ -239,6 +239,153 @@ Note that the form
+
+### RHS with expression list
+
+Instead of a failure-returning method call on the RHS of the statement, the RHS can instead be a list of expressions. As for a `:=` statement, in this form, there must be the same number of expressions on the left and right sides of `:-`. The semantics is very similar to that in the previous subsection. 
+   * The first RHS expression must have a failure-compatible type.
+   * All the assignments of RHS expressions to LHS values except for the first value are made.
+   * If the first RHS value (say `r`) responds `true` to `r.IsFailure()`, then `r.PropagateFailure()` is assigned to the first output value of the _caller_ and the execution of the caller's body is ended.
+   * If the first RHS value (say `r`) responds `false` to `r.IsFailure()`, then `r.Extract()` is assigned to the first LHS value of the `:-` statement. Execution of the caller's body continues with the statement following the `:-` statement.
+
+### Types without Extract
+
+An alternative form of the above behavior allows the failure-compatible type to omit the `Extract` method, A typical example would be
+```
+datatype VoidOutcome =
+| Success
+| Failure(error: string)
+{
+  predicate method IsFailure() {
+    this.Failure?
+  }
+  function method PropagateFailure(): VoidOutcome
+    requires IsFailure()
+  {
+  Failure(this.error); // this is Outcome<U>.Failure(...)
+  }
+}
+```
+Such a type also counts as a failure-compatible type for the purposes of a `:-` statement. However, now there is no `Extract` method and no value carried along with `Success`.
+The same exceptional return semantics apply. The only difference is that the list of LHS values now must omit what would be the first l-value, as there is no extracted value for it to get. Thus there may be no LHS values at all.
+
+In current Dafny, this form of exceptional return requires that the caller and callee each have only one output value, namely a value of this special failure-compatible extract-omitting type.
+
+TODO: Say more when the above restriction is lifted.
+
+### Chaining calls with potential failure
+
+Routinely using a first output value that has a failure-compatible type allows a programming style like that shown in the following example, using the `Outcome` class shown above.
+
+```
+method open(i: int) return (r: Outcome<int>) {
+  if (i < 0) return Failure("negative argument to a");
+  return Success(i+i);
+}
+
+method b(i: int) return (r: Outcome<bool>) {
+  if (i == 0) return Failure(zero argument to b");
+  return Success(100/i);
+}
+
+method c(i: int) return (r: Outcome<string>) {
+  if (i > 100) return Failure(big argument to c");
+  if (i > 0) Success("positive");
+  if (i < 0) Success("negative");
+  if (i == 0) Success("zero");
+}
+
+method m(i: int) return (r: Outcome<bool>){
+  var aa: int := a(i); // returns if i < 0
+  var bb: bool := b(i); // returns if i == 0
+  var cc: string := c(i); // returns if i > 100
```

Comment:
These lines should use `:-`, not `:=`.

Please run this whole example through Dafny to make sure there are no typos or other errors.

Reply:
Removed the example